### ISSUE TITLE
Stack overflow when running herald-wearable on nrf5340dk

### ISSUE THREAD

**Issue by liammcnair-fdh** [vmware/herald-for-cpp#50]
I am experiencing a stack overflow when running the sample project herald-wearable on the nrf5340DK. The only changes made to the clean repository is to remove security in the configuration and comment it out in the main.c. I then comment out all of the code blocks marked "PRODUCTION ONLY" and uncomment all the blocks marked "TESTING ONLY".

The way I can fix this is by adding the following configuration option:
`CONFIG_BT_RX_STACK_SIZE=8192`

**Comment by adamfowleruk** [comment]
Thanks for this. I'll be making a lot of memory management enhancements over the next few days. I'll change this setting to that you suggest for now and then use this story/issue to mark it as something to retest/verify as not required later.

**Comment by liammcnair-fdh** [comment]
Sounds great. Any memory management enhancements would be VERY exciting to us.

**Comment by adamfowleruk** [comment]
We've started rolling out the compile-time templates based approach in the herald::analysis API. Check out the following in-progress dev branch. We'll do similar things to the main API once this is proven and we've measured it. We believe that we can get up to 60% reduction in memory use by doing this, and allow wearable developers to configure a known maximum memory usage by component when building the code.

- analysis code: https://github.com/adamfowleruk/herald-for-cpp/tree/feature-42/herald/include/herald/analysis
- tests: https://github.com/adamfowleruk/herald-for-cpp/blob/feature-42/herald-tests/sample-tests.cpp and https://github.com/adamfowleruk/herald-for-cpp/blob/feature-42/herald-tests/analysisrunner-tests.cpp

I've also got a script to auto build the 'production' level hex's for all 4 DK hardware environments and package them. That should make it easier for people to test and verify.

**Comment by adamfowleruk** [comment]
@liammcnair-fdh I now have the memory use down so the demo apps will fit on all the test boards. See feature-42 branch in the above. I now have an issue as seen in #24 affecting the build on some systems including nRF52832dk. I've reported it to Nordic. If you could have a quick look too I'd appreciate it. It's a Zephyr core linker issue weirdly. For expected memory use see #24. It's not fantastic but I've only just started modifying the code for memory use. The Analysis API is extremely lean and so I'm going to apply those approaches to the rest of the code base, and rip out all inheritance and smart pointer use.

**Comment by liammcnair-fdh** [comment]
Amazing! I will spend some time Monday morning trying this out.

**Comment by adamfowleruk** [comment]
I've done lots of work on the code. Tests now seem to reliably pass but there are some odd issues still in the code. best to look at the feature-42 branch. Could you please see if your tools can locate the issue? I've drastically increased heap/pool sizes on the nRF5340 but I'm now getting this error somewhere in the OS code:-

```sh

00> [00:00:14.366,607] <dbg> heraldlogger.log: engine,coordinator,Sensor has Provider implementation
00> [00:00:14.366,821] <err> os: ***** USAGE FAULT *****
00> [00:00:14.366,821] <err> os:   Illegal use of the EPSR
00> [00:00:14.366,821] <err> os: r0/a1:  0x2000c0cc  r1/a2:  0x00000000  r2/a3:  0x2000bf30
00> [00:00:14.366,821] <err> os: r3/a4:  0x2000bf48 r12/ip:  0xc7ff75fb r14/lr:  0x00036fcf
00> [00:00:14.366,821] <err> os:  xpsr:  0x60000000
00> [00:00:14.366,851] <err> os: Faulting instruction address (r15/pc): 0x00000000
00> [00:00:14.366,851] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
00> [00:00:14.366,851] <err> os: Current thread: 0x20000db0 (unknown)
00> [00:00:14.703,094] <err> fatal_error: Resetting system
00> [00:00:00.364,654] <dbg> app.main: Logging test
```

These instructions seem to be in the Zephyr Bluetooth receiver code itself. I suspect there is a all to one of the C event handlers we have in the receiver code, and for some reason that memory has been stomped on.

I've been trying to use Ozone to debug this, but the UI is horrifically unfriendly and doesn't seem to find all the library sources successfully.

If you could please have a look I'd appreciate it. I'll check in my latest branch now and tag this issue. You should see a commit link appear below.

**Comment by adamfowleruk** [comment]
Ah I've got it working! There's a bugette in my new analysis code. I've disabled it for now on the feature-42 branch so you can try out the rest of the app tomorrow. It seems to work fine now.

You might also recognise the state who this detected device belongs to too ;) 👎 
```sh
00> [00:00:44.019,317] <dbg> heraldlogger.log: Sensor,BLE.ConcreteBLEReceiver,Finished reading CHAR read payload:-
00> [00:00:44.019,622] <dbg> heraldlogger.log: Sensor,BLE.ConcreteBLEReceiver,917c00300065005400414e657966674458736e344131374a2b414e657966674458736e344131374a2b414e657966674458736e344131374a2b414e6579666744~
00> [00:00:44.019,958] <dbg> app.sensor: sensor didRead: f1419ca09e46 with payload: 917c00300065005400414e657966674458736e344131374a2b414e657966674458736e344131374a2b414e657966674458736e344131374a2b414e657966674458736e344131374a2b414e657966674~
00> [00:00:44.019,989] <dbg> app.sensor: sensor didMeasure withPayload
```

I'll check in the code now. Let me know how it looks! I've got large pool/heap sizes on nRF5340 (well below the limit), but minimal ones on nRF52. You may need to play with them. Latest commit will appear below...

**Comment by adamfowleruk** [comment]
@liammcnair-fdh Also FYI I've just configured VSCode to run the debugger remotely with JLink. Took me 5 milliseconds to find the problem then! I'd highly recommend VSCode and the 'gnu-debugger' v1.0.4 JLink extension. Sample config files are already set up in that repo's launch config.

**Comment by adamfowleruk** [comment]
@liammcnair-fdh Now I have VSCode running with the nRF5340DK it's MUCH easier to debug these things!

I've found the source of all my woes. It's in the Context and Logging implementation. A mix of references, templates, and shared pointer code. I'm going to refactor the logging subsystem and context system. This will give us a rock solid base and also improve debug logging. No need for you to delve in to it.

If you could please try building the current feature-42 branch and deploying that locally, then try the other DKs (E.g. 52832) you have that'd be great. You may have to put your build in Release mode to fit it onto the 52832 right now. Please let me know how you get on.

Over the next couple of days I'll complete my logging system rewrite for Templates and we should be in good shape.

**Comment by liammcnair-fdh** [comment]
This is all great! I'll be back in the office tomorrow and try everything out. I'll let you know how it goes.

**Comment by liammcnair-fdh** [comment]
I have it all working in the Release build. See my implementation [here](https://github.com/liammcnair-fdh/hearald-wearable) that uses the feature-42 branch.

**Comment by adamfowleruk** [comment]
Great stuff. I'll close this issue for now. We can reopen it if the issue reappears.
Fixed by #42
